File: scripts/library-exchange.mjs
// SPDX-License-Identifier: Apache-2.0
// Actual wire interchange companion to check-parity.py; public synthetic fixtures only.
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import * as core from "@psp-cdl/core";
import * as crypto from "@psp-cdl/core/crypto";
import * as cdl from "@psp-cdl/cdl";
const fixture=name=>JSON.parse(readFileSync(new URL("../conformance/vectors/"+name,import.meta.url),"utf8"));
const codec=fixture("codec/profile-1.0.json"), signatures=fixture("signatures/profile-2.0.json");
const seed=Buffer.from(signatures.testKeys.ed25519.seedHex,"hex");
const {value,contentType,...metadata}=signatures.vectors[0].envelope.signature;
const inputs=[...codec.markupCases.map(c=>[c.id,core.parseMarkup(c.source)]),...codec.objectCases.map(c=>[c.id,core.documentFromObject(c.object)])];
const publicKey=e=>Buffer.from(e.signature.algorithm==="ed25519"?signatures.testKeys.ed25519.publicKeyHex:signatures.testKeys.hmac.keyHex,"hex");
function emitBundle(){
  return {
    documents:inputs.map(([id,doc])=>({id,markup:core.serializeMarkup(doc),canonical:core.serializeMarkup(doc,"canonical"),json:core.documentToJson(doc),semantic:core.documentToObject(doc,false)})),
    envelopes:[...signatures.vectors.map(v=>[v.id,crypto.signEnvelope(v.envelope.data,(({value,...m})=>m)(v.envelope.signature),v.envelope.signature.algorithm==="ed25519"?seed:publicKey(v.envelope))]),
      ["nested-document",crypto.signDocument(inputs.find(([id])=>id==="nested-application")[1],metadata,seed)]].map(([id,e])=>({id,json:core.serializeEnvelope(e),markup:core.serializeMarkup({kind:"document",children:[core.envelopeToSection(e)]}),inputHex:Buffer.from(core.signatureInput(e)).toString("hex")})),
    schemas:codec.cdlSchemas.map(c=>cdl.serializeCdlJson(c.schema)),
    declarations:["string","array"].map(format=>cdl.withDeclarations({title:"Portable declarations"},{classes:["PII"],covenants:["no-persist","no-training"],capabilities:[]},format))
  };
}
if(process.argv[2]==="--emit") process.stdout.write(JSON.stringify(emitBundle()));
else if(process.argv[2]==="--verify") {
  const peer=JSON.parse(readFileSync(0,"utf8"));
  assert.deepEqual(peer,emitBundle(),"Python serialization differs from TypeScript");
  for(const c of peer.documents){
    const fromJson=core.documentFromJson(c.json);
    assert.equal(core.serializeMarkup(fromJson),c.markup);
    assert.equal(core.serializeMarkup(fromJson,"canonical"),c.canonical);
    assert.deepEqual(core.documentToObject(core.parseMarkup(c.markup),false),c.semantic);
    assert.deepEqual(core.documentToObject(core.parseMarkup(c.canonical),false),c.semantic);
  }
  for(const c of peer.envelopes){
    const e=core.sectionToEnvelope(core.parseMarkup(c.markup).children[0]);
    assert.equal(core.serializeEnvelope(e),c.json);
    assert.equal(Buffer.from(core.signatureInput(e)).toString("hex"),c.inputHex);
    assert(crypto.verifySignature(e,publicKey(e)));
    if(c.id==="nested-document") assert.equal(core.serializeMarkup(core.envelopeToDocument(e)),codec.markupCases.find(c=>c.id==="nested-application").source);
  }
  for(const schema of peer.schemas) assert.equal(cdl.serializeCdlJson(cdl.parseCdlJson(schema)),schema);
  for(const d of peer.declarations) assert.deepEqual(cdl.parseDeclarations(d),{classes:["pii"],covenants:["no-persist","no-training"],capabilities:[]});
  console.log(`${peer.documents.length} documents and ${peer.envelopes.length} signed envelopes decoded from Python.`);
} else throw new Error("Use --emit or --verify");

File: scripts/library_exchange.py
# SPDX-License-Identifier: Apache-2.0
"""Independent Python producer/consumer for two-way library interoperability."""
import json
from pathlib import Path
import psp_cdl_core as core
from psp_cdl_core import crypto
import psp_cdl_cdl as cdl

ROOT = Path(__file__).resolve().parents[1]
def fixture(name):
    return json.loads((ROOT / 'conformance/vectors' / name).read_text(encoding='utf-8'))
CODEC = fixture('codec/profile-1.0.json')
SIGNATURES = fixture('signatures/profile-2.0.json')
SEED = bytes.fromhex(SIGNATURES['testKeys']['ed25519']['seedHex'])
INPUTS = [(c['id'], core.parse_markup(c['source'])) for c in CODEC['markupCases']] + [(c['id'], core.document_from_object(c['object'])) for c in CODEC['objectCases']]
def public_key(envelope):
    return bytes.fromhex(SIGNATURES['testKeys']['ed25519']['publicKeyHex'] if envelope['signature']['algorithm'] == 'ed25519' else SIGNATURES['testKeys']['hmac']['keyHex'])

def emit_bundle():
    metadata = {k:v for k,v in SIGNATURES['vectors'][0]['envelope']['signature'].items() if k not in {'value','contentType'}}
    envelopes = []
    for vector in SIGNATURES['vectors']:
        e = vector['envelope']
        envelopes.append((vector['id'], crypto.sign_envelope(e['data'], {k:v for k,v in e['signature'].items() if k != 'value'}, SEED if e['signature']['algorithm'] == 'ed25519' else public_key(e))))
    envelopes.append(('nested-document', crypto.sign_document(dict(INPUTS)['nested-application'], metadata, SEED)))
    return {
        'documents': [{'id': id, 'markup': core.serialize_markup(doc), 'canonical': core.serialize_markup(doc, 'canonical'), 'json': core.document_to_json(doc), 'semantic': core.document_to_object(doc, False)} for id,doc in INPUTS],
        'envelopes': [{'id': id, 'json': core.serialize_envelope(e), 'markup': core.serialize_markup({'kind':'document','children':[core.envelope_to_section(e)]}), 'inputHex': core.signature_input(e).hex()} for id,e in envelopes],
        'schemas': [cdl.serialize_cdl_json(c['schema']) for c in CODEC['cdlSchemas']],
        'declarations': [cdl.with_declarations({'title':'Portable declarations'}, {'classes':['PII'],'covenants':['no-persist','no-training'],'capabilities':[]}, format) for format in ('string','array')]
    }

def verify_bundle(peer):
    assert peer == emit_bundle(), 'TypeScript serialization differs from Python'
    for c in peer['documents']:
        document = core.document_from_json(c['json'])
        assert core.serialize_markup(document) == c['markup']
        assert core.serialize_markup(document, 'canonical') == c['canonical']
        assert core.document_to_object(core.parse_markup(c['markup']), False) == c['semantic']
        assert core.document_to_object(core.parse_markup(c['canonical']), False) == c['semantic']
    for c in peer['envelopes']:
        e = core.section_to_envelope(core.parse_markup(c['markup'])['children'][0])
        assert core.serialize_envelope(e) == c['json']
        assert core.signature_input(e).hex() == c['inputHex']
        assert crypto.verify_signature(e, public_key(e))
        if c['id'] == 'nested-document':
            assert core.serialize_markup(core.envelope_to_document(e)) == next(c['source'] for c in CODEC['markupCases'] if c['id']=='nested-application')
    for schema in peer['schemas']:
        assert cdl.serialize_cdl_json(cdl.parse_cdl_json(schema)) == schema
    for d in peer['declarations']:
        assert cdl.parse_declarations(d) == {'classes':['pii'],'covenants':['no-persist','no-training'],'capabilities':[]}
